Forge.TreeWalker.UnitTests/test/TreeSchemaValidatorTests.cs

```diff
@@ -0,0 +1,215 @@
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaInPath(schemaPath, stringRules, false, out List<string> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemaFromPathWithCustomRulesInJSchema()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaInPath(schemaPath, jschemaRules, false, out List<string> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemasFromDirectoryListWithCustomRulesInString()
+        {
+            bool res = ForgeSchemaValidator.ValidateMultipleSchemasInPath(schemaDirectoryPath, stringRules, false, out List<string> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemasFromDirectoryListWithCustomRulesInJSchema()
+        {
+            bool res = ForgeSchemaValidator.ValidateMultipleSchemasInPath(schemaDirectoryPath, jschemaRules, false, out List<string> errorList);
+            Assert.AreEqual(true, res); 
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateInvalidSchemaInString()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaString(stringSchema, jschemaRules, false, out List<string> errorList);
+            Assert.AreEqual(true, res);
+            Assert.AreEqual(0, errorList.Count);
+        }
+
+        [TestMethod]
+        public void Test_ValidateInvalidSchemaDirectoryPath()
+        {
+            bool res = ForgeSchemaValidator.ValidateMultipleSchemasInPath(invalidSchemaDirectoryPath, jschemaRules, false, out List<string> errorList);
+            Assert.AreEqual(false, res);
+            Assert.AreEqual("The directory name is invalid.\r\n", errorList.First());
+        }
+
+        [TestMethod]
+        public void Test_ValidateSchemaWithErrorContent()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaString(invalidSchemaWithErrorContent, jschemaRules, false, out List<string> errorList);
+            Assert.AreEqual(false, res);
+            Assert.AreEqual("JSON is valid against no schemas from 'oneOf'. line: 3, position: 17", errorList.First());
+        }
+
+        [TestMethod]
+        public void Test_ValidateInvalidSchemaAsDictionary()
+        {
+            bool res = ForgeSchemaValidator.ValidateSchemaString(invalidSchemaWithErrorContent, jschemaRules, true, out List<string> errorList);
+            Assert.AreEqual(false, res);
+            Assert.AreEqual("JSON is valid against no schemas from 'oneOf'. line: 3, position: 17", errorList.First());
+        }
+
+        [TestMethod]
+        public void Test_ValidateValidSchemaAsDictionary()
```

Comment:
>Test_ValidateValidSchemaAsDictionary [](http://example.com/codeflow?start=20&length=36)

The SubroutineSchema.json is a valid Dictionary<string, ForgeTree>. You can use that to test for dictionaries.

Reply:
Yea we aren't actually testing any dictionaries because we don't have a dictionary rules file. Let's add a basic one to contracts so we can test. I believe the below should work, but please test to confirm.

We can name the file: ForgeSchemaDictionaryValidationRules.json

{
  "type": "object",
  "patternProperties": {
    ".*?": {
      "$ref": "#/definitions/ForgeTreeDefinition"
    }
  },
  "additionalProperties": false,
  "definitions": {
    "ForgeTreeDefinition": {
      "type": "object",
      "properties": {
        "Tree": {
          "$ref": "[file://ForgeSchemaValidationRules.json#/definitions/TreeDefinition](file://ForgeSchemaValidationRules.json#/definitions/TreeDefinition)"
        },
        "RootTreeNodeKey": {
          "type": "string"
        }
      },
      "additionalProperties": false
    }
  }
}

Reply:
I think the ForgeSchemaDictionaryValidationRules can be exactly copy/pasted from what I put above. The point is for the ForgTreeDictionary rules to add one additional layer on top of the original rules, and then link to the original rules.

We can then use GetLinkedJSchemaRules to validate schemas that use dictionary.

Reply:
Hmm, the ForgeSchemaDictionaryVlaidationRules.json is a little different in iteration 7 than what I have above. Can you explain why there is a difference? Let's sync up offline.